pkg/translate/translate.go

```diff
@@ -14,34 +15,89 @@ type TranslateServiceServer struct {
 	pb.UnimplementedTranslateServiceServer
 }
 
+// ----------------------UploadTranslationFile-------------------------------
+
+type UploadParams struct {
+	Language LanguageData
+	Data     []byte
+	Schema   pb.Schema
+}
+
+// Validates request parameters for UploadTranslationFile.
+func (u *UploadParams) validate() error {
+	if len(u.Language.Str) == 0 {
+		return fmt.Errorf("'language' is required")
+	}
+
+	if len(u.Data) == 0 {
+		return fmt.Errorf("'data' is required")
+	}
+
+	var err error
+	if u.Language.Tag, err = language.Parse(u.Language.Str); err != nil {
+		return fmt.Errorf("parsing language '%s': %w", u.Language.Str, err)
+	}
+
+	return nil
+}
+
 func (t *TranslateServiceServer) UploadTranslationFile(
 	ctx context.Context,
 	req *pb.UploadTranslationFileRequest,
 ) (*emptypb.Empty, error) {
-	var (
-		reqLanguage = req.GetLanguage()
-		reqData     = req.GetData()
-		reqSchema   = req.GetSchema()
-	)
+	reqParams := UploadParams{
+		Language: LanguageData{Str: req.GetLanguage()},
```

Comment:
There are at least two steps involved in processing a request:

- parse request (type/schema guard)
- validate request (logical guard)
- access control
- etc

If you separate these steps, it simplifies the flow and removes unnecessary temporary data (E.g. Language struct with Str and Tag fields) that leaks and generates noise.

```go
// one of the options is...

type uploadParams struct {
  Language language.Tag
  // ...
}

params, err := parseUploadParams(req *pb.UploadTranslationFileRequest) (*uploadParams, error)

// ...

if err := params.validate(); err != nil {
```